Extend SubsetSpec API

jannes opened the pull request:
# Description
This closes #5024

extend [options.gloo.solo.io.SubsetSpec](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/subset_spec.proto.sk/#subsetspec) with:

- fallbackPolicy: string
- defaultSubset: []options.gloo.solo.io.Selector

extend [options.gloo.solo.io.Selector](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options/subset_spec.proto.sk/#selector) with:
- single_host_per_subset: boolean

These changes just expose already existing options of the Envoy API.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5024

kdorosh commented on projects/gloo/pkg/translator/clusters.go:
```diff
@@ -172,6 +175,33 @@ func createOutlierDetectionConfig(upstream *v1.Upstream) (*envoy_config_cluster_
 }
 
+func convertDefaultSubset(defaultSubset *v1_options.Subset) *_struct.Struct {
+	if defaultSubset != nil {
```
nit, it's easier to read with the code flattened

e.g.
```golang
if defaultSubset == nil {
  return nil
}

// rest of logic here
```

jannes replied:
thanks, i improved this

kdorosh commented:
/test thanks for the pr!

jannes commented:
I have modified one test to make sure the defaults are correct and added one test to check the defaultSubset translation. Let me know if you think there should be more tests:)

saiskee commented:
/test

saiskee commented on projects/gloo/api/v1/options/subset_spec.proto:
```diff
@@ -7,9 +7,24 @@ option (extproto.equal_all) = true;
+}
+
+enum FallbackPolicy {
+  // ANY_ENDOINT is the default
```
[super nit] `ANY_ENDPOINT`

saiskee commented on projects/gloo/api/v1/options/subset_spec.proto:
```diff
@@ -7,9 +7,24 @@ option (extproto.equal_all) = true;
 
 message SubsetSpec {
-    repeated Selector selectors =1;
+  repeated Selector selectors = 1;
```
Can we include comments describing each of these fields and the expected behavior? Any comments directly above the field will pulled into our official documentation when you run `make generated-code -B`.

jannes replied:
fixed

sam-heilbron commented:
Looks great! Letting it run through CI

sam-heilbron commented:
@jannes Our codegen has changed recently, which is why the codegen check is failing. Can you run `make install-go-tools generated-code -B` and then re-push?

sam-heilbron commented:
/test

jannes commented:
thanks for fixing the generated code and merging! :)
